src/metis/http/gist.lua

```diff
@@ -0,0 +1,285 @@
+    progress("Connecting to api.github.com... ")
+    local handle = http.get("https://api.github.com/gists/" .. id)
+    if handle == nil then
+        progress("Failed.\n")
+        return nil, "Failed to connect"
+    end
+    local meta = textutils.unserializeJSON(handle.readAll())
+    local code = handle.getResponseCode()
+    handle.close()
+    if code ~= 200 then
+        progress("Failed.\n")
+        return nil, "Invalid response code (" .. code .. ")" .. (meta and ": " .. meta.message or "")
+    end
+    if meta == nil or meta.files == nil then
+        progress("Failed.\n")
+        return nil, meta and "GitHub API error: " .. meta.message or "Error parsing JSON"
+    end
+    progress("Success.\n")
+    if file then return getGistFile(meta.files[file]), file
+    elseif next(meta.files, next(meta.files)) == nil then return getGistFile(meta.files[next(meta.files)]), next(meta.files)
+    elseif meta.files["init.lua"] ~= nil then return getGistFile(meta.files["init.lua"]), "init.lua"
+    else
+        local luaFiles = {}
+        for k in pairs(meta.files) do if k:match("%.lua$") then table.insert(luaFiles, k) end end
+        table.sort(luaFiles)
+        if #luaFiles == 0 then
+            progress("Error: Could not find any Lua files to download!\n")
+            return nil, "Could not find any Lua files to download"
+        end
+        if #luaFiles > 1 then progress("Warning: More than one Lua file detected, downloading the first one alphabetically.\n") end
+        return getGistFile(meta.files[luaFiles[1]]), luaFiles[1]
+    end
+end
+
+--- Runs a specified Gist. This is a wrapper for convenience.
+-- @tparam string id The Gist ID to download from. See above comments for more details.
+-- @tparam[opt] function progress A function to use to report status messages. If
+-- this is not a function, it will be used as an argument to the script.
+-- @tparam[opt] any ... Any arguments to pass to the script.
+-- @treturn any Any results returned from the script.
+function gist.run(id, progress, ...)
+    expect(1, id, "string")
+    local args = table.pack(...)
+    if type(progress) ~= "function" and progress ~= nil then
+        table.insert(args, 1, progress)
+        progress = nil
+    end
+    local data, name = gist.get(id, progress)
+    if data == nil then return end
+    local fn, err = load(data, name, "t", _ENV)
+    if fn == nil then error(err) end
+    local retval = table.pack(pcall(fn, table.unpack(args)))
+    if not retval[1] then error(retval[2]) end
+    return table.unpack(retval, 2)
+end
+
+--- Retrieves a table of all files from a Gist.
+-- @tparam string id The Gist ID to download.
+-- @tparam[opt] function progress A function to use to report status messages.
+-- @treturn table|nil A key-value list of all files in the Gist, or nil on error.
+-- @treturn string|nil If an error occurred, a string describing the error.
+function gist.getAll(id, progress)
+    expect(1, id, "string")
```

Comment:
This is very much duplicated from `get`. I wonder if it's worth providing a few more low-level functions like `parseId`, `getInfo`, etc...

I'd be inclined to leave the actual ID stripping code out of functions, and just make the caller do it - most of the time they shouldn't need to